app/services/chat_channels/create_invitation_link.rb

```diff
@@ -0,0 +1,15 @@
+module ChatChannels
+  module CreateInvitationLink
+    module_function
+
+    def call(chat_channel)
+      invitation_slug = "invitation-link-#{SecureRandom.hex(3)}"
+      chat_channel.update(invitation_slug: invitation_slug)
+      unless chat_channel.errors.any?
+        path = "/chat_channel_memberships/join_channel_invitation/#{chat_channel.slug}?invitation_slug=#{invitation_slug}"
```

Comment:
since the `invitation_slug` is in the path that we are going to use as an invite we dont need to store it in the database. When the link is clicked we can take the `invitation_slug` param, check if it is in Redis and if it is, complete the invite and if it is not, then we know it was expired.

Reply:
@mstruve 
`app/controllers/chat_channel_memberships_controller.rb`
`line no. 27`
We need to find the path to send as a response to the frontend. In that case, we don't have a `path`. So `invitation_slug` is required to save in DB. If I don't have the `invitation_slug`, It's not possible to find the path from Redis.

In this method, 1st we try to find the path using the invitation_slug. If invitation slug is not present in DB. That means there no path saved in Redis. In that case, we create the path and save it in Redis.

**Case 1** If the path is not present in the Redis `or` no Invitation slug, then we create a path and save on Redis, Also save the `invitation_slug` in DB. 
**Case 2** If we have the `invitation_slug` in DB. We read the REDIS with the `invitation_slug`. If path present in Redis. then pass to response else create a path in Redis with new `invitation_slug`

Reply:
> 1st we try to find the path using the invitation_slug. If invitation slug is not present in DB. That means there no path saved in Redis. In that case, we create the path and save it in Redis.

Rather than checking the DB, simply check Redis, if there is no path in Redis then that means the previous invite has expired and we need to generate a new one. If there is a path in Redis then we simply read that.

Reply:
@mstruve, We need the path with the unique `invitation_slug`. So if we don't know the `lnvitation_sug`. we cannot search on Redis.

**Reason**
  - If our path doesn't have a unique invitation_slug. It will be the same when you generate a new record in Redis. In that case if someone has the old URL/path after 12 hours he uses that link. He can use it to join the chat Channel. Because after 12 hours a new record will have an old path. 

That's why it is required to use the invitation_slug. It will change every 12 hours when the old expire.

Reply:
Here is what I am thinking
```ruby
def chat_channel_info
    @membership = ChatChannelMembership.find(params[:id])
    authorize @membership
    @channel = @membership.chat_channel
    
    invite_cache_key = "chat-channel-invite-#{chat_channel.id}" 
    invitation_slug = Rails.cache.fetch(invite_cache_key, expires_in: 12.hours) do 
        "invitation-link-#{SecureRandom.hex(3)}"
      end
    @invitation_link = "/join_channel_invitation/#{chat_channel.slug}?invitation_slug=#{invitation_slug}"

    if result&.errors&.any?
      render json: { success: false, errors: result.errors.full_messages }, status: :bad_request
    end
  end
```

The slug is stored in Redis and we fetch it using the chat channel ID. I do this using a handy little helper called [`fetch`](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch). This will read the cache key if it is present. If it is not present it will generate a new cache key. 

Once this is in place the next step is reading it when someone tries to use the link. 
```ruby
def join_channel_invitation
    @chat_channel = ChatChannel.find_by(slug: params[:channel_slug])
    authorize @chat_channel
    
    invite_cache_key = "chat-channel-invite-#{chat_channel.id}"  # could pull this out into a method
    invitation_slug = Rails.cache.read(invite_cache_key)
    
    if invitation_slug == params[:invitation_slug]
      # add them to the channel
    else
      # Show an error message that the invitation link has expired.
    end
  end
```

Let me know what you think!!!

Reply:
I got your point. Looks fine to me. I will update my code according to that.
 I was trying to save the full path on Radis.